runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,116 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
             TR::Node *destOffset = node->getChild(2);
             TR::Node *len = node->getChild(3);
             TR::Node *byteValue = node->getChild(4);
-            dest = TR::Node::create(TR::aladd, 2, dest, destOffset);
 
-            TR::Node * copyMemNode = TR::Node::createWithSymRef(TR::arrayset, 3, 3, dest, len, byteValue, node->getSymbolReference());
-            copyMemNode->setByteCodeInfo(node->getByteCodeInfo());
+            // When using balanced GC policy with offheap allocation enabled, there are three possible cases:
+            //
+            // 1.) The object at dest is known to be a non-array object at compile time. In this scenario, the final destination address
+            //     can be calculated right away as dest += destOffset. The IL node passed in to setmemoryEvaluator will have three
+            //     children: dest, len, and byteValue.
+            // 2.) The object at dest is known to be an array at compile time. In this scenario, we must load dataAddr from dest
+            //     to account for the way arrays are allocated when offheap is enabled. Once this adjustment is made, the final
+            //     destination address can be calculated dest += destOffset. Similar to case (1), the IL node passed in to
+            //     setmemoryEvaluator() will have three children: dest, len, and byteValue.
+            // 3.) The type of the object at dest is unknown at compile time. In this scenario, a runtime array check must be generated to
+            //     determine whether dataAddr needs to be loaded before calculating the final destination address. This means that we
+            //     cannot determine the final destination address right away, and thus need to keep dest and destOffset separate. In this
+            //     case, the node passed in to setmemoryEvaluator() will have four children: dest, destOffset, len, and byteValue.
+            //
+            // Note that if the default (gencon) GC policy is being used OR if the object is a NULL reference, dataAddr will never need
+            // to be loaded no matter the type of the object at dest, so all of the above cases will be treated like case (1).
+
+            //if offheap is not enabled OR if dest is known to be a NULL reference at compile time, default to case (1):
+            // - do NOT to load dataAddr to calculate final destination address
+            // - calculate the final destination address right away and only pass three children into setmemoryEvaluator()
+            bool loadDataAddr = false;
+            bool separateDestAndOffset = false;
+
+         #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+            if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit() && (!dest->isNull()))
```

Comment:
you have to understand this fact: compile-time flag is always conservative on the right side, i.e.  if ` isNull()`  were true, `dest` must be NULL;  if ` isNull()`  were false, `dest` could or could not be NULL. By and large, ` isNull()` is false from statistic perspective. Now, think about using it like this at this point:  for off-heap, it most likely goes to 3-children case (and without loading dataAddr -- since loadDataAddr is false at this point -- and you don't know if `dest`   is an array or not), it obviously is wrong.

Reply:
maybe you need to have a proper design before diving into implementation ... have all cases in front of you and flow-chart its implementation first.